src/cunumeric/index/choose_template.inl

```diff
@@ -0,0 +1,68 @@
+/* Copyright 2021 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "cunumeric/pitches.h"
+
+namespace cunumeric {
+
+using namespace Legion;
+using namespace legate;
+
+template <VariantKind KIND, LegateTypeCode CODE, int DIM>
+struct ChooseImplBody;
+
+template <VariantKind KIND>
+struct ChooseImpl {
+  template <LegateTypeCode CODE, int DIM>
+  void operator()(ChooseArgs& args) const
+  {
+    using VAL     = legate_type_of<CODE>;
+    auto out_rect = args.out.shape<DIM>();
+
+    Pitches<DIM - 1> pitches;
+    size_t volume = pitches.flatten(out_rect);
+    if (volume == 0) return;
+
+    auto out        = args.out.write_accessor<VAL, DIM>(out_rect);
+    auto index_rect = args.inputs[0].shape<DIM>();
+    auto index_arr  = args.inputs[0].read_accessor<int, DIM>(index_rect);
```

Comment:
The python side allows both 32 and 64 bit integers, but here you assume 32-bit integers. You likely want to assume 64-bit integers instead, and ensure this is the case (through casting) before calling to C++.